mint/run/core/versioning/list.go

```diff
@@ -0,0 +1,650 @@
+	}
+
+	if !reflect.DeepEqual(expectedResult, gotResult) {
+		failureLog(function, args, startTime, "", "ListObjectVersions returned unexpected listing result", nil).Fatal()
+		return
+	}
+
+	successLogger(function, args, startTime).Info()
+}
+
+// Test if version id marker continuation works in listing works well
+func testListObjectVersionsVersionIDContinuation() {
+	startTime := time.Now()
+	function := "testListObjectVersionIDContinuation"
+	bucket := randString(60, rand.NewSource(time.Now().UnixNano()), "versioning-test-")
+	object := "testObject"
+	expiry := 1 * time.Minute
+	args := map[string]interface{}{
+		"bucketName": bucket,
+		"objectName": object,
+		"expiry":     expiry,
+	}
+
+	_, err := s3Client.CreateBucket(&s3.CreateBucketInput{
+		Bucket: aws.String(bucket),
+	})
+	if err != nil {
+		failureLog(function, args, startTime, "", "CreateBucket failed", err).Fatal()
+		return
+	}
+	defer cleanupBucket(bucket, function, args, startTime)
+
+	putVersioningInput := &s3.PutBucketVersioningInput{
+		Bucket: aws.String(bucket),
+		VersioningConfiguration: &s3.VersioningConfiguration{
+			Status: aws.String("Enabled"),
+		},
+	}
+
+	_, err = s3Client.PutBucketVersioning(putVersioningInput)
+	if err != nil {
+		failureLog(function, args, startTime, "", "Put versioning failed", err).Fatal()
+		return
+	}
+
+	for i := 0; i < 10; i++ {
+		putInput1 := &s3.PutObjectInput{
+			Body:   aws.ReadSeekCloser(strings.NewReader("my content 1")),
+			Bucket: aws.String(bucket),
+			Key:    aws.String("testobject"),
+		}
+		_, err = s3Client.PutObject(putInput1)
+		if err != nil {
+			failureLog(function, args, startTime, "", fmt.Sprintf("PUT expected to succeed but got %v", err), err).Fatal()
+			return
+		}
+	}
+
+	input := &s3.ListObjectVersionsInput{
+		Bucket:  aws.String(bucket),
+		MaxKeys: aws.Int64(5),
+	}
+
+	type resultPage struct {
+		versions            []string
+		nextVersionIDMarker string
+		lastPage            bool
+	}
+
+	var gotResult []resultPage
+	var gotNextVersionIDMarker string
+	var numPages int
+
+	err = s3Client.ListObjectVersionsPages(input,
+		func(page *s3.ListObjectVersionsOutput, lastPage bool) bool {
+			numPages++
+			resultPage := resultPage{lastPage: lastPage}
+			if page.NextVersionIdMarker != nil {
+				resultPage.nextVersionIDMarker = *page.NextVersionIdMarker
+			}
+			for _, v := range page.Versions {
+				resultPage.versions = append(resultPage.versions, *v.Key)
+			}
+			if !lastPage {
+				// There is only two pages, so here we are saving the version id
+				// of the last element in the first page of listing
+				gotNextVersionIDMarker = *(*page.Versions[len(page.Versions)-1]).VersionId
+			}
+			gotResult = append(gotResult, resultPage)
+			return true
+		})
+
+	if err != nil {
+		failureLog(function, args, startTime, "", fmt.Sprintf("ListObjectVersions expected to succeed but got %v", err), err).Fatal()
+		return
+	}
+
+	if numPages != 2 {
```

Comment:
Relying on specific paging is beyond the S3 spec. The doc explicitly states that "less results may be returned".

Reply:
Instead it should check if the objects are returned as expected.

Reply:
> Relying on specific paging is beyond the S3 spec. The doc explicitly states that "less results may be returned".

I am not sure I understand here. Are you saying that the S3 server is free to return any number of items less than MaxKeys parameter ?

Reply:
This is the API promise:

> max-keys
> Sets the maximum number of keys returned in the response. By default the API returns up to 1,000 key names. The response might contain fewer keys but will never contain more. If additional keys satisfy the search criteria, but were not returned because max-keys was exceeded, the response contains <isTruncated>true</isTruncated>. To return the additional keys, see key-marker and version-id-marker.

Reply:
For me, we should not rely too much on their wording.

Is this test failing with any of your PRs ? if yes, does it only happen with versioning marker continuation ?

Reply:
@vadmeste I hacked around it, but tests shouldn't rely on implementation specifics, makes it less useful and will cause failures on valid responses.

Reply:
https://github.com/minio/minio/pull/11562 for reference.

Reply:
Exact paging is important here because we know how many objects we uploaded @klauspost - we know how much a server will return - max-keys=2 when the server has 2 keys it will return 2 keys not less. 

These tests are already tested against AWS S3 - if any of these tests fail then we have a bug.

Reply:
The response is only truncated if the prefix has more entries than max-keys, the doc states that even if you set 1000 if the server doesn't have 1000 then it will return less. I have never seen AWS return 900 objects when in-fact it has 1000 to return. 

Also these docs should be treated with a grain of salt, these are not something they publish for others to follow.  We have to behave from the point of view of the tests that work AWS S3 and not work with MinIO then we have a problem. 

There are more scenarios that still needs to be added which I will update this PR with.